src/magic_folder/config.py

```diff
@@ -451,31 +750,90 @@ def store_local_snapshot(self, cursor, snapshot):
 
         :param LocalSnapshot snapshot: The snapshot to store.
         """
-        # insert a new row or update an existing row with the new blob.
         try:
+            # Create the primary row.
             cursor.execute(
                 """
                 INSERT INTO
-                    [local_snapshots] ([path], [snapshot_blob])
+                    [local_snapshots] ([identifier], [path], [content_path])
                 VALUES
-                    (?, ?)
+                    (?, ?, ?)
                 """,
-                (snapshot.name, snapshot.to_json()),
+                (unicode(snapshot.identifier), snapshot.name, snapshot.content_path.path),
             )
         except sqlite3.IntegrityError:
-            # There is already a row with the given path.  Once we can depend
-            # on a newer SQLite3 we can use an UPSERT instead.  Meanwhile,
-            cursor.execute(
-                """
-                UPDATE
-                    [local_snapshots]
-                SET
-                    [snapshot_blob] = ?
-                WHERE
-                    [path] = ?
-                """,
-                (snapshot.to_json(), snapshot.name),
-            )
+            # The UNIQUE constraint on `identifier` failed - which *should*
+            # mean we already have a row representing this snapshot.
+            raise LocalSnapshotCollision(snapshot.identifier)
+
+        # Associate all of the metadata with it.
+        cursor.executemany(
+            """
+            INSERT INTO
+                [local_snapshot_metadata] ([snapshot_identifier], [key], [value])
+            VALUES
+                (?, ?, ?)
+            """,
+            list(
+                (unicode(snapshot.identifier), k, v)
+                for (k, v)
+                in snapshot.metadata.items()
+            ),
+        )
+        # Associate all of the parents with it.  First remote, then local.
+        #
+        # XXX What is the relative ordering of parents_local and
+        # parents_remote?  It's not preserved in `LocalSnapshot`.  Since local
+        # parents turn into remote parents I guess I'll put remote parents
+        # first here.  That way as local snapshots disappear they'll disappear
+        # from the middle and appear at the end of the remote snapshots list,
+        # also in the middle?
```

Comment:
I don't believe we should be depending upon the *order* of parents anywhere. In the vast majority of cases there will be just 1 parent .. so it'll flow from local to remote. Only when there are conflicts will there be 2+ parents .. but they don't have any "order".

Reply:
Okay.  This would have been good to figure out earlier in the process, I guess.  I preserved order in the database because the Python API preserves order and there was never a good argument articulated (that I saw) that order doesn't matter.

Since this code works, I'm going to leave it (because it's easier to do so at this point).  In the future, if order doesn't matter, we can get rid of all of the ordering support in the database layer *and* change the Python representation so it doesn't appear to represent an order (probably switch some lists to some sets).